node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,46 +61,379 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
+									<SubsystemFieldInitMethod::<#var_type>>(boxed_func),
+								spawner: self.spawner,
+							}
+						}
+					}
+				};
+				setter.extend(setter_with);
 
-	let consumes = &info.consumes();
+				// For subsystems, we also need a replacement method
+				let setter_name_replace = format_ident!("replace_{}", setter_name);
+				let modified_generic_ty = quote!{
+					OverseerFieldInit<SubsystemFieldInitMethod<NEW>>
+				};
+				let mut modified_generic_return_type = return_type_generics.clone();
+				modified_generic_return_type[idx] = &modified_generic_ty;
+				// We can use idx here as it is exactly the same as subsystems_wihout_wip idx
+				let cur_consumes = &consumes[idx];
+				let new_where_condition = quote! {
+					#support_crate ::Subsystem<#subsystem_ctx_name< #cur_consumes >, #error_ty>
+				};
+				// Another note is that replace function requires the subsystem field
+				// to be initialized first
+				let setter_replace = quote! {
+					impl <S, #( #impl_generics, )*>
+					#builder <S, #( #return_type_generics, )*> #builder_where_clause {
+						/// Replace a subsystem by another implementation for the
+						/// consumable message type.
+						pub fn #setter_name_replace<NEW, F>(self, gen_replacement_fn: F)
+							-> #builder <S, #( #modified_generic_return_type, )*>
+						where
+							#var_type: 'static,
+							F: 'static + FnOnce(#var_type) -> NEW,
+							NEW: #new_where_condition,
+						{
+							let replacement: SubsystemFieldInitMethod<NEW> = match self.#setter_name.0 {
+								SubsystemFieldInitMethod::Fn(fx) =>
+									SubsystemFieldInitMethod::Fn(Box::new(move |handle: #handle| {
+									let orig = fx(handle)?;
+									Ok(gen_replacement_fn(orig))
+								})),
+								SubsystemFieldInitMethod::Value(val) =>
+									SubsystemFieldInitMethod::Value(gen_replacement_fn(val)),
+							};
+							#builder::<S, #( #modified_generic_return_type, )*> {
+								#setter_name: OverseerFieldInit::
+									<SubsystemFieldInitMethod::<NEW>>(replacement),
+								#(
+									#other_fields_definitions,
+								)*
+								spawner: self.spawner,
+							}
+						}
+					}
+				};
+				setter.extend(setter_replace);
+			}
+			setter
+		})
+		.collect::<Vec<_>>();
 
-	let subsyste_ctx_name =
-		Ident::new(&(overseer_name.to_string() + "SubsystemContext"), overseer_name.span());
+	// Used as a implementation spec for a builder with all fields uninitialised
+	let all_uninit_generics = field_helpers
+		.iter()
+		.map(|fh| &fh.uninited_ty)
+		.collect::<Vec<_>>();
+	// Similarly, defines all field types as initialised
+	let all_init_generics = field_helpers
+		.iter()
+		.map(|fh| &fh.inited_ty)
+		.collect::<Vec<_>>();
+	// Defines all fields to be set to uninit<T>
+	let uninit_builder_field_def = field_helpers
+		.iter()
+		.map(|fh| {
+			let (field_name, field_ty) = (&fh.name, &fh.field_ty);
+			if fh.is_subsystem {
+				quote! { #field_name: OverseerFieldUninit::<SubsystemFieldInitMethod<#field_ty>>::default() }
```

Comment:
Why is there a nesting here? Could we not make this one level and make `OverseerFieldInit` an `enum`?